fix(broker): fix lost messages on rejected correlation

MiguelPires opened the pull request:
Sends a reset command which removes the message's correlation from the state, allowing it to be correlated again.

closes #1811

saig0 commented on broker-core/src/main/java/io/zeebe/broker/subscription/message/processor/ResetMessageCorrelationProcessor.java:
```diff
@@ -0,0 +1,81 @@
+          .setCorrelationKey(subscription.getCorrelationKey())
+          .setMessageName(subscription.getMessageName());
+
+      streamWriter.appendNewCommand(MessageSubscriptionIntent.CORRELATE, correlateMessageRecord);
```
If the subscription is closed then we don't write an event as confirmation of the command. This is very uncommon for our stream processing.

MiguelPires commented:
Hey @saig0, I refactored the reject command and added the test case you mentioned. As we discussed, I opened an issue for the second case, where messages appear to be correlated out of order https://github.com/zeebe-io/zeebe/issues/2459

saig0 commented on broker-core/src/main/java/io/zeebe/broker/subscription/message/processor/ResetMessageCorrelationProcessor.java:
```diff
@@ -62,20 +70,30 @@ public void processRecord(
 
-      streamWriter.appendNewCommand(MessageSubscriptionIntent.CORRELATE, correlateMessageRecord);
-    }
+    messageCorrelator.correlateNextMessage(subscription, correlateMessageRecord, sideEffect);
```
This looks a bit odd because we already have the message to correlate with ;) I would like to have it more explicitly to be sure that we correlate the rejected message.

saig0 commented on broker-core/src/main/java/io/zeebe/broker/subscription/message/processor/MessageEventProcessors.java:
```diff
@@ -77,7 +77,8 @@ public static void addMessageProcessors(
             MessageSubscriptionIntent.RESET,
-            new ResetMessageCorrelationProcessor(messageState, subscriptionState))
+            new ResetMessageCorrelationProcessor(
+                messageState, subscriptionState, subscriptionCommandSender))
```
We should also change the intent to "REJECT" to be aligned with the subscription command.

saig0 commented on broker-core/src/main/java/io/zeebe/broker/subscription/message/processor/ResetMessageCorrelationProcessor.java:
```diff
@@ -62,20 +70,30 @@ public void processRecord(
     }
     messageState.removeMessageCorrelation(messageKey, workflowInstanceKey);
 
-    final MessageSubscription subscription =
```
We should write an "ACKNOWLEDGE" event for the command (with intent "REJECTED"). Otherwise, we would have no follow-up event if no subscription exists.

saig0 commented on broker-core/src/test/java/io/zeebe/broker/subscription/message/MessageStreamProcessorTest.java:
```diff
@@ -375,6 +375,53 @@ public void shouldCorrelateMultipleMessagesTwoBeforeOpen() {
+            captor.capture(),
+            eq(messageKey),
+            any(DirectBuffer.class));
+    assertThat(BufferUtil.equals(captor.getValue(), secondSubscription.getMessageName())).isTrue();
```
Why do we need to verify the message name? It is equal in both cases.

saig0 commented on broker-core/src/test/java/io/zeebe/broker/workflow/message/MessageCorrelationTest.java:
```diff
@@ -678,6 +681,59 @@ public void shouldCorrelateToNonInterruptingBoundaryEvent() {
   }
 
+  @Test
+  public void shouldCorrelateBothMessages() {
```
```suggestion
  public void shouldCorrelateMessageAgainAfterRejection() {
```

MiguelPires commented:
Hey @saig0, I iterated again on the rejection processor

MiguelPires commented:
bors r+

MiguelPires commented:
bors retry
